src/poetry/installation/executor.py

```diff
@@ -158,11 +160,37 @@ def execute(self, operations: list[Operation]) -> int:
                     operation.package.develop
                     and operation.package.source_type in {"directory", "git"}
                 )
-                if not operation.skipped and is_parallel_unsafe:
+                # Skipped operations are safe to execute in parallel
+                if operation.skipped:
+                    is_parallel_unsafe = False
+
+                if is_parallel_unsafe:
                     serial_operations.append(operation)
-                    continue
+                elif operation.package.source_type == "git":
+                    # Git operations on the same repository should be executed serially
+                    serial_git_operations[operation.package.source_url].append(
+                        operation
+                    )
+                else:
+                    tasks.append(
+                        self._executor.submit(self._execute_operation, operation)
+                    )
 
-                tasks.append(self._executor.submit(self._execute_operation, operation))
+            # For each git repository, execute all operations serially
+            for repository_git_operations in serial_git_operations.values():
+
+                def _serialize(
+                    repository_serial_operations: list[Operation],
+                ) -> None:
+                    for operation in repository_serial_operations:
+                        self._execute_operation(operation)
+
```

Comment:
Minor nit: given that you chose to pass the serial ops as a parameter, can `_serialize` be moved out of the loop?

Reply:
I'm sorry for being unclear in the above comment. I meant: move `_serialize` out of the loop, but still inside the `execute` function.

I am not convinced that making `_serialize` a member function provides value. It's a helper function very closely coupled with the parallel execution happening in `execute`.